src/main/java/org/maxgamer/maxbans/transaction/TransactionLayer.java

```diff
@@ -0,0 +1,75 @@
+package org.maxgamer.maxbans.transaction;
+
+import org.hibernate.Session;
+import org.hibernate.SessionFactory;
+import org.hibernate.Transaction;
+import org.hibernate.resource.transaction.spi.TransactionStatus;
+import org.maxgamer.maxbans.exception.TransactionException;
+
+import java.io.Closeable;
+
+/**
+ * Represents a single transaction layer. The root transaction layer is responsible for committing
+ * the final result, where any nested transactions are disregarded when closed.
+ */
+public class TransactionLayer implements Closeable {
+    private ThreadLocal<Session> sessions;
+    private Session session;
+    private Transaction transaction;
+
+    private boolean opened;
+
+    public TransactionLayer(ThreadLocal<Session> sessions, SessionFactory factory) {
```

Comment:
Instead of `TransactionLayer` (Creating many objects for each possible Transaction), could I use a single `Tx` object, and have `transact()` return the same `Tx` object, which just increments a counter?

Pros:
* Memory saving
* Access to the parent transaction / Tx

Cons:
* Mishandling a subtransaction (forgetting to call close() or use a try-catch) would now destroy the entire root transaction